library/core/src/panic.rs

```diff
@@ -62,6 +63,68 @@ pub macro panic_2021 {
     }),
 }
 
+#[doc(hidden)]
+#[unstable(feature = "edition_panic", issue = "none", reason = "use panic!() instead")]
+#[allow_internal_unstable(
+    core_panic,
+    core_intrinsics,
+    const_dispatch,
+    const_eval_select,
+    const_format_args,
+    panic_args,
+    rustc_attrs
+)]
+#[rustc_diagnostic_item = "core_panic_2021_macro"]
+#[rustc_macro_transparency = "semitransparent"]
+#[cfg(not(any(bootstrap, feature = "panic_immediate_abort")))]
+pub macro panic_2021 {
+    () => ({
+        // Create a function so that the argument for `track_caller`
+        // can be moved inside if possible.
+        #[cold]
+        #[track_caller]
+        #[inline(never)]
+        const fn panic_cold_explicit() -> ! {
+            $crate::panicking::panic_explicit()
+        }
+        panic_cold_explicit();
+    }),
+    // Special-case the single-argument case for const_panic.
+    ("{}", $arg:expr $(,)?) => ({
+        #[cold]
+        #[track_caller]
+        #[inline(never)]
+        #[rustc_do_not_const_check] // Allow the call to `panic_display`.
```

Comment:
`panic_display` is a `const fn`, so why is this attribute needed?

Reply:
It's constrained to only work with `&'static str` by const checking, so we can't pass the argument of type `&T` to it.

Reply:
Ah... that's un-pretty.

I think the better approach would be to generalize what happens to `panic_display`. I'm imagining an attribute like `#[const_panic_display]` such that all functions with this attribute
- are allowed to be called in const context if the argument type is `&str`
- if evaluation reaches them, the magic hook kicks in

That way, both the original `fn panic_display` and the outlined functions you are adding here can just get that attribute, and we don't need all this `rustc_do_not_const_check` and `const_eval_select` stuff here.